File: python-alert-service/models/triggered_alert.py
"""Pydantic models for triggered alert data."""

from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class AlertStatus(str, Enum):
    """Valid triggered alert statuses."""

    OPEN = "open"
    ACKNOWLEDGED = "acknowledged"
    RESOLVED = "resolved"


class TriggeredAlert(BaseModel):
    """Complete triggered alert model."""

    id: str = Field(..., description="Unique alert identifier")
    rule_id: str = Field(..., description="ID of the rule that triggered this alert")
    sensor_id: str = Field(..., description="ID of the sensor that triggered this alert")
    sensor_value: float = Field(..., description="Sensor value at time of trigger")
    threshold: float = Field(..., description="Threshold that was crossed")
    message: str = Field(..., description="Human-readable alert message")
    status: AlertStatus = Field(..., description="Alert status")
    created_at: str = Field(..., description="When the alert was triggered")
    resolved_at: Optional[str] = Field(None, description="When the alert was resolved (None until resolved)")

    class Config:
        from_attributes = True


class TriggeredAlertUpdate(BaseModel):
    """Model for updating a triggered alert status."""

    status: AlertStatus = Field(..., description="New alert status")


class TriggeredAlertList(BaseModel):
    """Response model for list of triggered alerts."""

    alerts: list[TriggeredAlert]
    count: int

File: go-alert-service/models/triggered_alert.go
package models

import "fmt"

// ValidAlertStatuses contains all allowed triggered alert statuses.
var ValidAlertStatuses = map[string]bool{
	"open":         true,
	"acknowledged": true,
	"resolved":     true,
}

// TriggeredAlert represents a complete triggered alert record from the database.
type TriggeredAlert struct {
	ID          string  `json:"id"`
	RuleID      string  `json:"rule_id"`
	SensorID    string  `json:"sensor_id"`
	SensorValue float64 `json:"sensor_value"`
	Threshold   float64 `json:"threshold"`
	Message     string  `json:"message"`
	Status      string  `json:"status"`
	CreatedAt   string  `json:"created_at"`
	ResolvedAt  *string `json:"resolved_at"`
}

// TriggeredAlertUpdate represents the request body for updating a triggered alert.
type TriggeredAlertUpdate struct {
	Status *string `json:"status" binding:"required"`
}

// Validate checks if the TriggeredAlertUpdate fields are valid.
func (u *TriggeredAlertUpdate) Validate() error {
	if u.Status != nil && !ValidAlertStatuses[*u.Status] {
		return fmt.Errorf("invalid alert status: %s", *u.Status)
	}
	return nil
}

// TriggeredAlertList represents the response for listing triggered alerts.
type TriggeredAlertList struct {
	Alerts []TriggeredAlert `json:"alerts"`
	Count  int              `json:"count"`
}

// HealthResponse represents the health check response.
type HealthResponse struct {
	Status  string `json:"status"`
	Service string `json:"service"`
}

// ErrorResponse represents an error response.
type ErrorResponse struct {
	Detail string `json:"detail"`
}
